hhs_facilities/delphi_hhs_facilities/generate_signals.py

```diff
@@ -0,0 +1,56 @@
+"""Functions for generating signals."""
+
+from typing import Callable
+
+import pandas as pd
+import numpy as np
+
+
+def generate_signal(df: pd.DataFrame,
+                    input_cols: list,
+                    signal_func: Callable,
+                    date_offset: int) -> pd.DataFrame:
+    """
+    Generate a signal DataFrame derived from an input DataFrame.
+
+    Applies the provided function the the columns specified, and then aggregates by geo and time.
+
+    Parameters
+    ----------
+    df: pd.DataFrame
+        Input DataFrame containing columns specified in `input_cols`.
+    input_cols: list of strings
+        List of column names to pass to `signal_func`.
+    signal_func: function
+        Function which takes in a list of Series and produces a signal Series.
+    date_offset: integer
+        Number of days to add to the timestamp. This is used because some of the columns are
+        "previous_day_<metric>" and require us adding -1 days to accurate depict the right timespan.
+
+    Returns
+    -------
+    Signal DataFrame that is ready for `create_export_csv`.
+    """
+    df_cols = [df[i] for i in input_cols]
+    df["val"] = signal_func(df_cols)
+    df["timestamp"] = df["timestamp"] + pd.Timedelta(days=date_offset)
+    df = df.groupby(["timestamp", "geo_id"], as_index=False).sum()
+    df["se"] = np.nan
+    df["sample_size"] = np.nan
+    return df
+
+
+def sum_cols(cols: list) -> pd.Series:
+    """
+    Sum a list of Series, requiring 1 non-nan value per row sum.
+
+    Parameters
+    ----------
+    cols: list of Series
+        List of Series to sum.
+
+    Returns
+    -------
+    Series of summed inputs.
+    """
+    return pd.concat(cols, axis=1).sum(axis=1, min_count=1)
```

Comment:
if all values are nan, is the row removed from the output?

Reply:
Not from this function, since it needs to be the same dimension as the input df to become a new column. I can add a dropna() statement so we don't get this behavior
```
In [18]: df                        
Out[18]: 
   val geo
0  NaN   a
1  NaN   a
2  3.0   b
3  4.0   b

In [19]: df.groupby('geo', as_index=False).sum()                          
  geo  val
0   a  0.0
1   b  7.0
```

and instead get 
```
In [20]: df.dropna(subset=['val']).groupby('geo', as_index=False).sum()                    
Out[20]: 
  geo  val
0   b  7.0
```